nntrainer/layers/activation_layer.cpp

```diff
@@ -77,13 +72,12 @@ int ActivationLayer::setActivation(
 
 int ActivationLayer::setActivation(
   std::function<Tensor &(Tensor const &, Tensor &)> const &activation_fn,
-  std::function<Tensor &(Tensor const &, Tensor &)> const
-    &activation_prime_fn) {
+  std::function<Tensor &(Tensor &, Tensor &)> const &activation_prime_fn) {
   _act_fn = activation_fn;
-  _act_prime_fn = [activation_prime_fn](Tensor const &x, Tensor &ret_derivative,
+  _act_prime_fn = [activation_prime_fn](Tensor &x, Tensor &ret_derivative,
                                         Tensor const &derivative) -> Tensor & {
-    ret_derivative = activation_prime_fn(x, ret_derivative);
-    ret_derivative.multiply_i(derivative);
+    x = activation_prime_fn(x, x);
```

Comment:
Q. is this one okay when the optimization is off?

Reply:
Check the last commit which ensures that everything is tested with optimization on as well as off.
The reason this works is that all the layers work with their inputs in the backward operation.
This `x` is actually the input of the next layer which has already performed its backward operation (note that backward start from last layer to the first layer). So, overwriting this `x` does not affect other layers.

Reply:
Thanks for the confirmation :)

Reply:
When optimization is off, this shouldn't replace the backward result isn't it?

Reply:
When the optimization is off, still behaves the same as of now.
Layer's behavior must remain the same independent of the optimization.
The layer shouldn't know if the optimization is being done or not.

The activation layer now always works in-place with this. It is up to the model to use that information for optimization.

Reply:
Oh I C. Thanks for the explanation :)